crates/testing/src/block_builder.rs

```diff
@@ -279,3 +225,244 @@ pub fn run_random_builder(url: Url) {
+        _signature: &<<TestTypes as NodeType>::SignatureKey as SignatureKey>::PureAssembledSignatureType,
+    ) -> Result<AvailableBlockData<TestTypes>, BuildError> {
+        let mut blocks = self.blocks.write().await;
+        let entry = blocks.get_mut(block_hash).ok_or(BuildError::NotFound)?;
+        entry.payload.take().ok_or(BuildError::Missing)
+    }
+
+    async fn claim_block_header_input(
+        &self,
+        block_hash: &BuilderCommitment,
+        _signature: &<<TestTypes as NodeType>::SignatureKey as SignatureKey>::PureAssembledSignatureType,
+    ) -> Result<AvailableBlockHeaderInput<TestTypes>, BuildError> {
+        let mut blocks = self.blocks.write().await;
+        let entry = blocks.get_mut(block_hash).ok_or(BuildError::NotFound)?;
+        entry.header_input.take().ok_or(BuildError::Missing)
+    }
+}
+
+impl SimpleBuilderSource {
+    pub async fn run(self, url: Url) {
+        let builder_api =
+            hotshot_builder_api::builder::define_api::<SimpleBuilderSource, TestTypes, Version01>(
+                &Options::default(),
+            )
+            .expect("Failed to construct the builder API");
+        let mut app: App<SimpleBuilderSource, hotshot_builder_api::builder::Error, Version01> =
+            App::with_state(self);
+        app.register_module("/", builder_api)
+            .expect("Failed to register the builder API");
+
+        async_spawn(app.serve(url, STATIC_VER_0_1));
+    }
+}
+
+pub struct SimpleBuilderTask {
+    transactions: Arc<RwLock<HashMap<Commitment<TestTransaction>, TestTransaction>>>,
+    blocks: Arc<RwLock<HashMap<BuilderCommitment, BlockEntry>>>,
+}
+
+impl TaskState for SimpleBuilderTask {
+    type Event = Arc<HotShotEvent<TestTypes>>;
+
+    type Output = HotShotTaskCompleted;
+
+    fn filter(&self, event: &Arc<HotShotEvent<TestTypes>>) -> bool {
+        !matches!(
+            event.as_ref(),
+            HotShotEvent::TransactionsRecv(_)
+                | HotShotEvent::LeafDecided(_)
+                | HotShotEvent::Shutdown
+        )
+    }
+
+    async fn handle_event(
+        event: Self::Event,
+        task: &mut Task<Self>,
+    ) -> Option<HotShotTaskCompleted> {
+        let this = task.state_mut();
+        match event.as_ref() {
+            HotShotEvent::TransactionsRecv(transactions) => {
+                let mut queue = this.transactions.write().await;
+                for transaction in transactions {
+                    queue.insert(transaction.commit(), transaction.clone());
+                }
+            }
+            HotShotEvent::LeafDecided(leaf_chain) => {
+                let mut queue = this.transactions.write().await;
+                for leaf in leaf_chain.iter() {
+                    if let Some(ref payload) = leaf.block_payload {
+                        for txn in payload.transaction_commitments(&()) {
+                            queue.remove(&txn);
+                        }
+                    }
+                }
```

Comment:
This logic won't remove transactions if the `TransactionRecv` event happens after a leaf containing that transaction is decided.  It's pretty common with the webserver we download transactions in batches right before we become a leader.  So we'll get lots of transactions that were already decided.  I'm not sure how relevant that is for this builder if we only use as a test.  If we do bench-marking with it then we probably need to handle this case.